qutip/solver/mcsolve.py

```diff
@@ -385,6 +408,9 @@ def __init__(self, H, c_ops, *, options=None):
                 rhs -= 0.5 * n_op
 
         self._num_collapse = len(self._c_ops)
+        self.options = options
+        if self.options["improved_sampling"]:
+            self.resultclass = McResultImprovedSampling
```

Comment:
This is something I'm not super happy about: this wants to be defined outside of init (where we have ``resultclass = McResult`` now). Maybe still not a good enough reason to define a whole new class ``MCSolverImprovedSampling`` though

Reply:
Or maybe set `resultclass` to be a property that check the option and return the right type.

Another options would by using a metaclass. It feels overkill, but would be clean (here at least).

Reply:
The metaclass idea sounds fun but maybe the below simple fix works?